runner.py

```diff
@@ -1123,22 +1136,45 @@ def run(self):
                 self._debugger.pause('Remove phase complete. Waiting to stop and cleanup')
 
             self.end_measurement()
-            self.check_stderr_processes()
+            self.check_process_returncodes()
             self.custom_sleep(config['measurement']['idle-time-end'])
-            self.stop_metric_providers()
             self.store_phases()
             self.update_start_and_end_times()
         except BaseException as exc:
             self.add_to_log(exc.__class__.__name__, str(exc))
             raise exc
         finally:
             try:
-                self.read_stdout_logs()
-                self.read_and_cleanup_processes()
-                self.save_notes_runner()
-                self.save_stdout_logs()
+                self.read_container_logs()
+            except BaseException as exc:
+                self.add_to_log(exc.__class__.__name__, str(exc))
+                raise exc
             finally:
-                self.cleanup()  # always run cleanup automatically after each run
+                try:
+                    self.read_and_cleanup_processes()
+                except BaseException as exc:
+                    self.add_to_log(exc.__class__.__name__, str(exc))
+                    raise exc
+                finally:
+                    try:
+                        self.save_notes_runner()
+                    except BaseException as exc:
+                        self.add_to_log(exc.__class__.__name__, str(exc))
+                        raise exc
+                    finally:
+                        try:
+                            self.save_stdout_logs()
+                        except BaseException as exc:
+                            self.add_to_log(exc.__class__.__name__, str(exc))
+                            raise exc
+                        finally:
+                            try:
+                                self.stop_metric_providers()
+                            except BaseException as exc:
+                                self.add_to_log(exc.__class__.__name__, str(exc))
+                                raise exc
+                            finally:
+                                self.cleanup()  # always run cleanup automatically after each run
```

Comment:
Something like this might be clearer:
```
def execute_and_log_on_exception(method):
    try:
        method()
    except BaseException as exc:
        self.add_to_log(exc.__class__.__name__, str(exc))
        return exc
    return []

try:
    self.read_container_logs()
except BaseException as exc:
    self.add_to_log(exc.__class__.__name__, str(exc))
    raise exc
finally:
    raised = []
    raised += execute_and_log_on_exception(self.read_and_cleanup_processes)
    raised += execute_and_log_on_exception(self.save_notes_runner)
    raised += execute_and_log_on_exception(self.save_stdout_logs)
    raised += execute_and_log_on_exception(self.stop_metric_providers)
    self.cleanup()  # always run cleanup automatically after each run

    if raised:
        for rais in raised:
            print(f"raised: {rais}")
        raise BaseException("Some shit went down")
```

But it is more code and probably more verbose. I am not sure.